### PR TITLE
InfluxDB persistence for influxdb versions >= 0.9 

### PR DESCRIPTION
**Pull request by theoweiss**
This PR should replace PR https://github.com/openhab/openhab/pull/3322 .
All commits from #3322 are squashed into one.

### CODE REVIEW

**Comment by theoweiss** [comment]
@teichsta can you please make a brief check if the PR looks good.

File: bundles/persistence/org.openhab.persistence.influxdb/pom.xml [hunk 1]
```diff
@@ -12,7 +12,7 @@
 	<name>openHAB InfluxDB Persistence</name>
 
 	<properties>
-		<bundle.symbolicName>org.openhab.persistence.influxdb</bundle.symbolicName>
+		<bundle.symbolicName>org.openhab.persistence.influxdb1</bundle.symbolicName>
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/pom.xml]: revert this change back to .influxdb

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+
+  public void unsetItemRegistry(ItemRegistry itemRegistry) {
+    this.itemRegistry = null;
+  }
+
+  public void activate(final BundleContext bundleContext, final Map<String, Object> config) {
+    logger.debug("influxdb persistence service activated");
+    disconnect();
+    password = (String) config.get("password");
+    if (StringUtils.isBlank(password)) {
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: statically import isBlank() for the sake of readability

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+      isProperlyConfigured = false;
+      logger
+          .error("influxdb:password",
+              "The password is missing. To specify a password configure the password parameter in openhab.cfg.");
+      return;
+    }
+
+    url = (String) config.get("url");
+    if (StringUtils.isBlank(url)) {
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: statically import isBlank() for the sake of readability

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+      url = DEFAULT_URL;
+      logger.debug("using default url {}", DEFAULT_URL);
+    }
+
+    user = (String) config.get("user");
+    if (StringUtils.isBlank(user)) {
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: statically import isBlank() for the sake of readability

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+      user = DEFAULT_USER;
+      logger.debug("using default user {}", DEFAULT_USER);
+    }
+
+    dbName = (String) config.get("db");
+    if (StringUtils.isBlank(dbName)) {
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: statically import isBlank() for the sake of readability

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+          dbStatus = true;
+          logger.debug("database status is OK");
+        } else {
+          logger.error("database connection failed, version is: \"{}\" response time was \"{}\"",
+              version, pong.getResponseTime());
+          dbStatus = false;
+        }
+      } catch (RuntimeException e) {
+        dbStatus = false;
+        logger.error("database connection failed throwing an exception");
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: maybe add the e.getMessage() to the log statement?

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+      logger.error("checkConnection: database is not connected");
+      dbStatus = false;
+    } else {
+      try {
+        Pong pong = influxDB.ping();
+        String version = pong.getVersion();
+        // may be check for version >= 0.9
+        if (version != null && !version.contains("unknown")) {
+          dbStatus = true;
+          logger.debug("database status is OK");
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: you could also log the version here

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+        } else {
+          logger.error("database connection failed, version is: \"{}\" response time was \"{}\"",
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: is correct to log "connection failed"? At this point we can only say that the ping failed, right? A physical connection could be available though.

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+
+    State state = null;
+    if (item.getAcceptedCommandTypes().contains(HSBType.class)) {
+      state = item.getStateAs(HSBType.class);
+      logger.trace("got hsbtype", state.toString());
+    } else if (item.getAcceptedDataTypes().contains(PercentType.class)) {
+      state = item.getStateAs(PercentType.class);
+      logger.trace("percentype item found {}", name);
+    } else {
+      logger.trace("other item found");
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: better unify the different states to be logged here in order to ease debugging. Something linke "Tried to get item as {}"

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 10]
```diff
@@ -0,0 +1,582 @@
...
+      state = item.getState();
+    }
+    Object value = stateToObject(state);
+    logger.trace("storing {} in influxdb value {}, {}", name, value, item);
+    Point point =
+        Point.measurement(name).field(VALUE_COLUMN_NAME, value)
+            .time(System.currentTimeMillis(), timeUnit).build();
+    try {
+      influxDB.write(dbName, "default", point);
+      // influxDB.write(batchPoints);
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: remove commented code

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 11]
```diff
@@ -0,0 +1,582 @@
...
+      return Collections.emptyList();
+    }
+
+    List<HistoricItem> historicItems = new ArrayList<HistoricItem>();
+
+    StringBuffer query = new StringBuffer();
+    query.append("select ");
+    query.append(VALUE_COLUMN_NAME);
+    // query.append(", ");
+    // query.append(TIME_COLUMN_NAME);
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: remove commented code

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java [hunk 12]
```diff
@@ -0,0 +1,582 @@
...
+    }
+
+    logger.trace("filter itemname: {}", filter.getItemName());
+    logger.trace("filter ordering: {}", filter.getOrdering().toString());
+    logger.trace("filter state: {}", filter.getState());
+    logger.trace("filter operator: {}", filter.getOperator());
+    logger.trace("filter getBeginDate: {}", filter.getBeginDate());
+    logger.trace("filter getEndDate: {}", filter.getEndDate());
+    logger.trace("filter getPageSize: {}", filter.getPageSize());
+    logger.trace("filter getPageNumber: {}", filter.getPageNumber());
```
Comment (teichsta) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: better combine these statements into one

**Comment by teichsta** [comment]
> @teichsta can you please make a brief check if the PR looks good.

sure, please find my review comments inline. After incorporating the feedback, please squash and merge. Please update the wiki accordingly. It should be clear to the former users that this change is API breaking.

**Comment by sumnerboy12** [comment]
Great work @theoweiss - been waiting for this so I can upgrade my InfluxDB instance!

File: bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java (hunk 6 above)
Comment (theoweiss) [inline bundles/persistence/org.openhab.persistence.influxdb/src/main/java/org/openhab/persistence/influxdb/internal/InfluxDBPersistenceService.java]: I changed it this way:
```
  logger.error("database connection failed", e);
```

**Comment by theoweiss** [comment]
I incorporated your feedback and squashed the commits.
Now I'm testing the influxdb 0.9 service one more time.

**Comment by theoweiss** [comment]
I hesitate to merge this PR because:
it's really annoying but this "order by DESC" bug mentioned here https://github.com/openhab/openhab/issues/2748#issuecomment-153864938 changed, but still exists:
```
> select value from ATestTime where  time < 1451517300s ORDER BY time ASC
name: ATestTime
---------------
time		value
1451517180	1.451517120221e+12
1451517240	1.451517180085e+12

> select value from ATestTime where  time < 1451517300s ORDER BY time DESC
> 
```
I found this issue https://github.com/influxdb/influxdb/issues/4944 but it is still open.
I've deleted all influxdb data files and switched to tsm1 but nothing changed:
```
$ grep tsm /usr/local/etc/influxdb.conf.0.9.theo3
  dir = "/Users/theo/.influxdb/datatsm1"
  engine = "tsm1"
```
Any hints how to solve this?

**Comment by teichsta** [comment]
since this is a known (and documented) bug i would tend to merge this PR nevertheless in order to let the community test the new API (implementation). Unfortunately there doesn't seem to be that much drive on the issue. I might help to comment on the orig issue @theoweiss?

**Comment by theoweiss** [comment]
@teichsta I also thought about the best solution and came to the same decision: let's merge the PR and have the opportunity to get it fixed by the influxdb people or find a workaround for 1.8.1.
